tests/e2e/v3_lease_no_proxy_test.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/pkg/v3/expect"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"go.etcd.io/etcd/tests/v3/framework/testutils"
+)
+
+// TestLeaseRevokeByOldLeader verifies that leases shouldn't be revoked by
+// old leader.
+// See the first case in the section "Root Cause" in
+// https://docs.google.com/document/d/1peLDjwebnSuNR69ombuUiJ5fkSw7Rd3Slv_mUpOLsmg/edit
+func TestLeaseRevokeByOldLeader(t *testing.T) {
+	testLeaseRevokeIssue(t, func(cluster *e2e.EtcdProcessCluster, leaderIndex int) []string {
+		// return the endpoint of a follower
+		return cluster.Procs[(leaderIndex+1)%len(cluster.Procs)].EndpointsGRPC()
+	})
+}
+
+// TestLeaseRevokeByNewLeader verifies that leases shouldn't be revoked by
+// new leader.
+// See the second case in the section "Root Cause" in
+// https://docs.google.com/document/d/1peLDjwebnSuNR69ombuUiJ5fkSw7Rd3Slv_mUpOLsmg/edit
+func TestLeaseRevokeByNewLeader(t *testing.T) {
+	testLeaseRevokeIssue(t, func(cluster *e2e.EtcdProcessCluster, leaderIndex int) []string {
+		// return all members' endpoints, so that the client can switch to
+		// other member in case the leader gets stuck on writing disk.
+		return cluster.EndpointsGRPC()
+	})
+}
+
+func testLeaseRevokeIssue(t *testing.T, epsFn func(cluster *e2e.EtcdProcessCluster, leaderIndex int) []string) {
```

Comment:
Please don't use function as an argument. It externalizes important part of the test making it harder to track what exactly is tested. Instead you can just provide a bool to switch behavior within a test.

Reply:
Resolved although do not fully agree. Not a big deal though.